---
title: "Using crisprDesign to design gRNAs with minor and major alleles"
author: Jean-Philippe Fortin, Luke Hoberecht
output: 
  github_document:
    toc: true
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```


# Introduction

Genetic variants such as single nucleotide polymorphisms (SNPs) can 
be problematic in guide RNA (gRNA) design, as different alleles can
result in unintended gRNA:DNA mismatches for on-targets that reduce 
gRNA efficacy. To circumvent this, it is advisable to generally avoid
targeting sequences that contain variants. However, this may not always
be possible, due to a small target window and/or few target options,
or desirable, if, for example, a CRISPR application intends to target 
a pathogenic variant.

Functions in `crisprDesign` are well equipped to handle these cases.
gRNAs overlapping SNPs can be identified with the `addSNPAnnotation` 
function, as documented in the [CRISPRko design with Cas9](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) tutorial. Should the user wish to target region despite (or because of)
the presence of variants, the user only needs to take care in the choice
of `BSgenome` when constructing the `GuideSet` (an alternative option is
to supply a custom target sequence; see the [Working with a custom sequence](https://github.com/crisprVerse/Tutorials/tree/master/Design_Custom_Sequence) tutorial for more information).

This tutorial covers use cases for `BSgenome` objects that store variants 
of the reference human genome (hg38) injected with major and minor alleles.
It assumes the reader is familiar with constructing and using gene 
annotation objects (see the [Building a gene annotation object](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation) tutorial) and `GuideSet` objects (see the [CRISPRko design with Cas9]( https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) tutorial) so that the content may focus on the utility of 
the `BSgenome` variants discussed herein. Please consult the
applicable tutorials if necessary.

Finally, it goes without saying that the user should be knowledgeable
of the sequence(s), including possible variations in such, he or she 
is designing gRNAs for.



# Installation

See the [Installation tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Installation) to learn how to install the crisprVerse packages necessary
for this tutorial. 



# Terminology

See the [CRISPRko design tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar 
with the terminology used throughout this tutorial.

# gRNA design

## Loading packages

We first load the necessary packages for this tutorial:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprBase)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major)
library(BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor)
```

We will also load the gene annotation model `txdb_human` from the
`crisprDesignData` package:

```{r}
data(txdb_human, package="crisprDesignData")
```


The `BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major` and 
`BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor` packages are `BSgenome` 
packages that contain the major and minor alleles of the human reference
genome hg38 based dbSNP151.For more information, type the following: 


```{r, eval=FALSE}
help(BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major)
help(BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor)
```

## Designing gRNAs for human (hg38) with injected major alleles

It is worth noting that the human reference genome sequence (GRCh38.p12), 
does not give the major allele 
(i.e. most common allele in a population) at all nucleotide locations.
Indeed, given that it was historically constructed from a small set of 
human genomes, it contains minor alleles that were common across this 
set of human genomes.

For example, in the coding region (CDS) of the human gene SMC3, 
the reference genome contains the minor allele of the SNP rs2419565; the
reference allele (A) frequency is 0.00577, and the alternative allele (G) 
frequency is 0.99423 as indicated in the table here [here](https://www.ncbi.nlm.nih.gov/snp/rs2419565)). 


Designing gRNAs targeting the CDS of SMC3 with the SpCas9 nuclease returns
one gRNA that overlaps this SNP. Below, we first construct a 
`GuideSet` object using the reference `BSgenome` object:


```{r, collapse=TRUE, results='markup'}
smc3 <- queryTxObject(txdb_human,
                      featureType="cds",
                      queryColumn="gene_symbol",
                      queryValue="SMC3")
gs_reference <- findSpacers(smc3,
                            crisprNuclease=SpCas9,
                            bsgenome=BSgenome.Hsapiens.UCSC.hg38)
``` 

and a `GuideSet` object with the `BSgenome` object that contains 
the major alleles:

```{r, collapse=TRUE, results='markup'}
gs_major <- findSpacers(smc3,
                        crisprNuclease=SpCas9,
                        bsgenome=BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major)
```

Let's compare the protospacer sequences from both objects:

```{r, collapse=TRUE, results='markup'}

protospacers(gs_reference["spacer_199"])
protospacers(gs_major["spacer_199"])
```

The variant occurs in the seed sequence of this gRNA, 5 bases upstream of
the `pam_site`, so a gRNA:DNA mismatch at this location is likely detrimental
to its efficacy. Also, as this major allele occurs at >99% frequency, it
may be more beneficial to design gRNAs in this example using the major allele
genome contained in `BSgenome.Hsapiens.UCSC.hg38.dbSNP151.major`.


## Designing gRNAs for human (hg38) with injected minor alleles

It may be desirable, in some applications, to target a genic sequence 
that contains a minor allele (i.e. less common allele) rather than the 
major or reference allele. For example, if a particular minor allele is 
pathogenic and the host cell has a single copy of that allele, the user
may want to target that pathogenic variant and disrupt its behavior while
leaving the other copy undisturbed.

As an example, using `BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor`, we can 
target the pathogenic minor allele ([rs398122995](https://www.ncbi.nlm.nih.gov/clinvar/variation/92240/?oq=rs398122995&m=NM_001378454.1(ALMS1):c.1897C%3ET%20(p.Gln633Ter))) located in the human gene
ALMS1:



```{r}
alms1 <- queryTxObject(txdb_human, 'cds', 'gene_symbol', 'ALMS1')
gs_minor <- findSpacers(alms1,
                        crisprNuclease=SpCas9,
                        bsgenome=BSgenome.Hsapiens.UCSC.hg38.dbSNP151.minor)
gs_minor <- unique(gs_minor)
```

We also include, for comparison, the resulting `GuideSet` using the 
reference genome sequence:

```{r}
gs_reference <- findSpacers(alms1,
                            crisprNuclease=SpCas9,
                            bsgenome=BSgenome.Hsapiens.UCSC.hg38)
gs_reference <- unique(gs_reference)
```

and compare the two versions of the gRNA:

```{r}
gs_reference["spacer_615"]
gs_minor["spacer_612"]
```

The variant occurs 1 base upstream of the `pam_site`, and likely influences
gRNA activity, that is, we can design a gRNA that targets the minor allele 
and has a much lower affinity for the reference, or major allele.

Note that while the two `GuideSet` objects differ only by their
`BSgenome` object, we need to provide different indices to access
protospacers at equivalent `pam_site`s. This is due to variants in
one `BSgenome` (in this case theone with minor alleles) eliminating
PAM sequences, that is, one of the Gs  in NGG is changed to another base 
such that SpCas9 does not recognize it. This, where permissible, is also 
an effective way of ensuring gRNAs only  target a specific sequence if
that sequence contains the desired variant.


# Session Info

```{r}
sessionInfo()
```



