---
title: "Building a gRNA database for the human transcriptome using CasRx (RfxCas13d)"
output: 
  github_document:
    toc: true
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin

Date: July 16, 2022


# Introduction

In this tutorial, we provide reproducible code to design and annotates all gRNAs in a transcriptome, for a given organism, and for a DNA-targeting nuclease such as Cas13d. As an example, we design CRISPRkd gRNAs for the human transcriptome, in GRCh38 coordinates, for the  nuclease RfxCas13d (CasRx).

# Loading necessary packages

We first load the necessary packages:


```{r, warning=FALSE, message=FALSE}
library(crisprBase)
library(crisprScore)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
```


### Specifying the genome

We specify a `BSGenome` object that contains the reference genome of the target organism:

```{r}
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
```

### Specifying the genome index

We specify the file path of a Bowtie index build on a transcriptome (different from an index built on a reference genome). Briefly, the index was build on a fasta file storing mRNA sequences for a given gene model (here Ensembl release 104). See the Genome index tutorial to learn how to build such an index.  

```{r}
bowtie_index <- "/Users/fortinj2/crisprIndices/bowtie/ensembl_human_104/ensembl_human_104"
```


### Specifying the nuclease

The `crisprBase` provides functionalities for creating custom `crisprNuclease` objects, and also provides already-available `crisprNuclease` objects such as the commonly-used `SpCas9`, `enAsCas12a`, `CasRx` nucleases. 

Here, we will use the DNA-targeting nuclease `CasRx`:

```{r}
data(CasRx, package="crisprBase")
crisprNuclease=CasRx
```

### Specifying on-target scoring methods

We specify which on-target scoring methods we want to use:

```{r}
scoring_methods=c("casrxrf")
```

More information about which scoring methods are available for a given nuclease can be obtained using the following:


```{r}
crisprScore::scoringMethodsInfo
```


### Specifying gene models and TSS annotations

We also need a gene model formatted as a `TxDbObject`. The `crisprDesignData` contains such objects for both the human and mouse genomes, in GRCh38 (hg38) and GRCm38 (mm10) coordinates, respectively. Ensembl gene models were used to generate such objects.


```{r, warning=FALSE, message=FALSE}
data(txdb_human, package="crisprDesignData")
txObject <- txdb_human
```

The `crisprDesignData` vignette shows how to create such objects from other transcriptomes. 



# Building a complete annotation for a given gene 


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Explicitely calling those packages here instead of
# printout when calling precomputeGuides. This can be omitted. 
library(crisprScoreData)
library(ExperimentHub)
library(AnnotationHub)
library(BiocFileCache)
library(dbplyr)
```

The `designCompleteAnnotation` function in `crisprDesign` provides a one-step workflow to design and annotate gRNAs targeting a coding gene for a user-specific combination of parameters.

Here, we design all gRNAs targeting the main KRAS isoform (ENST00000311936) for CRISPRkd applications using CasRx:


```{r, warning=FALSE}
gs <- designCompleteAnnotation(queryValue="ENST00000311936",
                               queryColumn="tx_id",
                               modality="CRISPRkd",
                               bsgenome=bsgenome,
                               bowtie_index=bowtie_index,
                               crisprNuclease=CasRx,
                               txObject=txObject,
                               n_mismatches=1,
                               scoring_methods=scoring_methods)
```


The resulting object is a `GuideSet` object containing the fully annotated gRNAs, as described in the `crisprDesign` vignette. See the Tutorial `CRISPRkd` to learn more about `GuideSet` objects and how to access their rich annotations for RNA-targeting nucleases such as CasRx.


```{r}
gs
```


### Converting the `GuideSet` object to a list of data.frames

The `flattenGuideSet` function in `crisprDesign` is a convenience function to convert the complex `GuideSet` object into a set of 5 data.frames that can be saved as plain text files:

```{r}
dfs <- flattenGuideSet(gs)
```

We can look at the names of the data.frames:

```{r}
names(dfs)
```

As an example, let's look at the first rows of the primary data.frame:

```{r}
head(dfs$primary)
```

# Building a complete annotation for all genes

We first get all possibles transcripts from our gene model:

```{r}
tx_ids <- unique(txObject$cds$tx_id)
head(tx_ids)
```

and build an index to loop over for generating the GuideSet objects:

```{r}
tx_index <- seq_along(tx_ids)
head(tx_index)
```

We specify where to save the GuideSets:


```{r, eval=FALSE}
dir <- "./crisprko_casrx_hg38"
if (!dir.exists(dir)){
  dir.create(dir, recursive=TRUE)
}
```

We are now ready to generate and save all data with the function `designCompleteAnnotation` from `crisprDesign`. The function was designed to be as comprehensive as possible to design and annotate gRNAs in one step. It does the following: 

- Extract the DNA/RNA sequences with `queryTss`/`queryTxDB`
- Design gRNAs with `findSpacers`
- Remove gRNAs targeting repeat elements with `removeRepeats`
- Characterize spacer sequences with `addSequenceFeatures`
- Find on- and off-targets with `addSpacerAlignmentsIterative`
- Add gene annotation with `addGeneAnnotation`
- Add TSS annotation with `addTssAnnotation`
- Add on-target efficiency scores with `addOnTargetScores`
- Add off-target specificity scores with `addOffTargetScores`
- Add SNP annotation with `addSNPAnnotation`
- Add restriction enzymes information with `addRestrictionEnzymes`


We are now looping over all transcripts to generate the data:

```{r, eval=FALSE}
lapply(tx_index, function(tx){
    gs <- designCompleteAnnotation(queryValue=tx,
                                   queryColumn="tx_id",
                                   modality="CRISPRkd",
                                   bsgenome=bsgenome,
                                   bowtie_index=bowtie_index,
                                   crisprNuclease=CasRx,
                                   txObject=txObject,
                                   n_mismatches=3,
                                   scoring_methods=scoring_methods)
    write.rds(gs, file=file.path(dir, paste0(tx, ".rds")))
})
```

This loop can be modified by the user to use an embarrassingly-parallel approach to save time using the `BiocParallel` package, for instance.


# Reproducibility

```{r}
sessionInfo()
```
