---
title: "Validating existing gRNA libraries"
output: 
  github_document:
    toc: true
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin

Date: July 8, 2022


# Introduction

In this vignette, we characterize a small mouse CRISPR knockout (CRISPRko) library that was designed to target tumor suppressors. 
The library was obtained from Addgene, and is stored in `.\extdata`.

# Loading necessary packages

```{r, warning=FALSE, message=FALSE}
library(crisprDesign)
library(crisprBowtie)
library(crisprBase)
library(readxl)
library(BSgenome.Mmusculus.UCSC.mm10)
bsgenome <- BSgenome.Mmusculus.UCSC.mm10
```

We also load `crisprDesignData`, which is a data package containing
already-processed Ensembl objects for gene annotation of human and mouse gRNAs:

```{r, warning=FALSE, message=FALSE}
library(crisprDesignData)
```

# Reading in data

```{r}
data <- read_excel("extdata/mtsg-grnas-readcounts.xlsx")
data <- as.data.frame(data)[,1:2]
colnames(data) <- c("ID", "spacer_20mer")

# Getting genes names:
data$gene_symbol <- sapply(strsplit(data$ID, split="_"), function(x)x[[1]])
head(data)
```


# Building a `GuideSet` object


We first define the nuclease for the analysis. 
We here use the standard wildtype Cas9 (SpCas9) from
the `crisprBase` package:

```{r}
data(SpCas9, package="crisprBase")
crisprNuclease <- SpCas9
crisprNuclease
```

The default length of the spacer sequences is 20nt.
This can be changed to a different length if needed,
for instance 19nt:

```{r, eval=FALSE}
# Not sunr
spacerLength(SpCas9) <- 19
```

We next need to define a bowtie index that we will use for alignment:


```{r}
bowtie_index <- "/Users/fortinj2/crisprIndices/bowtie/mm10/mm10"
```


We first map the gRNAs to the reference genome with perfect match to obtain genomic coordinates:

```{r}
spacers <- unique(data$spacer_20mer)
aln <- runCrisprBowtie(spacers,
                       crisprNuclease=crisprNuclease,
                       bowtie_index=bowtie_index,
                       n_mismatches=0)
head(aln)
```

`n_mismatches=0` specifies that we require a perfect match between
spacer and protospacer sequences (on-targets).

Non-targeting controls should not have any alignments to the genome,
and some guides might have multiple alignments if they were not
designed carefully. For such guides, that's OK, we can pick up pick one 
genomic coordinate for now, and the multiple alignments annotation
will be handled later on. 

We keep only alignments to the standard chromosomes:

```{r}
chrs <- paste0("chr",c(1:22, "X", "Y"))
aln <- aln[aln$chr %in% chrs,,drop=FALSE]
```


We add the genomic coordinates to the data.frame:

```{r}
wh <- match(data$spacer_20mer, aln$spacer)
data$chr <- aln$chr[wh]
data$pam_site <- aln$pam_site[wh]
data$pam <- aln$pam[wh]
data$strand <- aln$strand[wh]
head(data)
```


We can now build a proper `GuideSet` object in `crisprDesign` that will 
allow us to do (more) sophisticated analyses.


We need to filter out first guides that don't have a match to the genome:

```{r}
data <- data[!is.na(data$pam_site),,drop=FALSE]
```

Ready to build the object with the constructor function in crisprDesign:

```{r}
gs <- GuideSet(protospacers=data$spacer_20mer,
               pams=data$pam,
               pam_site=data$pam_site,
               seqnames=data$chr,
               strand=data$strand,
               CrisprNuclease=crisprNuclease,
               bsgenome=bsgenome)
gs$gene_symbol <- data$gene_symbol
```

While it is not necessary, we will unique names to the spacers:

```{r}
names(gs) <- paste0("gRNA_", seq_along(gs))
```

The `GuideSet` object, and `crisprDesign`, provide rich functionalities to annotate and manipulate gRNAs; see the main `crisprDesign` vignette for more details. For the rest of this vignette, we only focus on characterizing the off-targets.


# Off-target characterization

Having a `GuideSet` object, it is now a piece of cake to characterize the off-targets. We characterize off-targets using the bowtie aligner, with up to 3 mismatches between the spacer (gRNA) and protospacer (target DNA) sequences.
The function `addSpacerAlignments` accomplishes that. 

It has an optional argument `txObject` that can be used to provide gene model data to put the off-targets in a gene model context. We made such objects available for human and mouse in the `crisprDesignData` package (see `txdb_human` and `txdb_mouse).




```{r, eval=TRUE, warning=FALSE}
data(txdb_mouse, package="crisprDesignData")
txObject <- txdb_mouse
gs <- addSpacerAlignments(gs,
                          txObject=txObject,
                          aligner="bowtie",
                          aligner_index=bowtie_index,
                          bsgenome=bsgenome,
                          n_mismatches=2)
```


The alignments are stored in a metadata column called `alignments`. See ?getSpacerAlignments for more details about what the different columns are. 

As an example, we can access the on- and off-target alignments of the first gRNA using the following:


```{r, eval=TRUE}
aln <- gs$alignments[[1]]
aln
```


We can also add CFD and MIT scores to the off-targets to characterize the likelihood of SpCas9 to cut at the off-targets:

```{r, eval=TRUE}
gs <- addOffTargetScores(gs)
```

The scores range from 0 to 1, and a higher score indicates a higher probability of the off-target to occur. 





# Session Info

```{r}
sessionInfo()
```
