---
title: "Using crisprDesign to design gRNAs that map across species"
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin, Luke Hoberecht

Date: `r format(Sys.time(), '%d %B, %Y')`


# Introduction

This tutorial describes how to design guide RNAs (gRNAs) that target homologous genes across multiple species using functions in the `crisprDesign` package. This strategy can be applied to any two (or more) species for which the genome sequence and gene model annotation is available.


# Installation

This tutorial used several packages from Bioconductor that can be
installed using the following commands:


```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("crisprDesign")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
BiocManager::install("BSgenome.Mmusculus.UCSC.mm10")
```

and the following package from GitHub:

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("Jfortin1/crisprDesignData")
```


# Terminology

See the [CRISPRko design tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar 
with the terminology used throughout this tutorial.


# Mapping gRNAs across species

## Loading packages

We first load the necessary packages for this tutorial:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprBase)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
library(BSgenome.Mmusculus.UCSC.mm10)
```


## Creating the GuideSet

In this tutorial we will design gRNAs using the SpCas9 nuclease that target both the human KRAS gene and the homologous mouse Kras gene. There are multiple ways to go about this, which we describe in the following sections. Each strategy, however, begins by constructing a `GuideSet` targeting a "host" gene. 

We will take the human KRAS gene as our "host" gene and create a `GuideSet` targeting the CDS of that gene. We first load the SpCas9 `CrisprNuclease` object from the `crisprBase` package and data containing gene regions for the human genome from the `crisprDesignData` package, `txdb_human` (we will also load a similar object for the mouse genome, `txdb_mouse`). For more information on `txdb_human` and `txdb_mouse` and how to create similar gene annotation objects, see the [Building a gene annotation object tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation) tutorial.

```{r}
data(SpCas9, package="crisprBase")
data(txdb_human, package="crisprDesignData")
data(txdb_mouse, package="crisprDesignData")
```

Next, we find the coordinates for the CDS of KRAS using the `queryTxObject` function then build our `GuideSet` with the `findSpacers` function.

```{r}
kras_human <- queryTxObject(txdb_human,
                            featureType="cds",
                            queryColumn="gene_symbol",
                            queryValue="KRAS")
gs_human <- findSpacers(kras_human,
                        crisprNuclease=SpCas9,
                        bsgenome=BSgenome.Hsapiens.UCSC.hg38) 
```


## Mapping gRNAs across species via `intersect`

For this strategy we first create a similar `GuideSet` targeting the mouse homolog Kras.

```{r}
kras_mouse <- queryTxObject(txdb_mouse,
                            featureType="cds",
                            queryColumn="gene_symbol",
                            queryValue="Kras")
gs_mouse <- findSpacers(kras_mouse,
                        crisprNuclease=SpCas9,
                        bsgenome=BSgenome.Mmusculus.UCSC.mm10) 
```

Then, we find the common spacers between the two `GuideSet`s using `intersect` and filter each for this common spacer set.

```{r}
common_spacers <- intersect(spacers(gs_human),
                            spacers(gs_mouse))
results_human <- gs_human[spacers(gs_human) %in% common_spacers]
results_human
results_mouse <- gs_mouse[spacers(gs_mouse) %in% common_spacers]
results_mouse
```

This simple approach, however, has some drawbacks. It requires gRNAs to have perfect sequence matching, which, while perhaps acceptable for targets having many gRNA choices, may be too restrictive for those applications that have fewer choices and may need tolerate mismatches in the target genes. Also, and more notably, we now have multiple `GuideSets` to maintain in the process of selecting candidate gRNAs (see [CRISPRko design with Cas9](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9))--essentially twice the work!


## Mapping gRNAs across species via `addSpacerAlignments`

To avoid the drawbacks of the above strategy, we can use the `addSpacerAlignments` function on our human KRAS `GuideSet` to append alignment annotation of the **mouse** genome.

In demonstrating this method, we will use the bowtie aligner; we need to specify a bowtie index for the mouse genome: 

```{r}
# Path of the mm10 bowtie index on my personal laptop:
bowtie_index_mouse <- "/Users/hoberecl/crisprIndices/bowtie/mm10/mm10"
# bowtie_index <- "/Users/fortinj2/crisprIndices/bowtie/mm10/mm10"
```

For instructions on how to build a Bowtie index from a given reference genome, see the [genome index tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Genome_Indices) or the [crisprBowtie page](https://github.com/Jfortin1/crisprBowtie).

We will also search up to 1 mismatch and pass the gene model object `txdb_mouse` to the `txObject` argument, so the alignments will be annotated with genomic context and we can determine which of our spacers map to the CDS of Kras.

As we will also want to search for off-targets in the human genome in a later step, we can ensure these results are not overwritten by setting the `colname` argument to a non-default value, such as `alignments_mouse`.

```{r}
results_human <- addSpacerAlignments(gs_human,
                                     aligner="bowtie",
                                     aligner_index=bowtie_index_mouse,
                                     bsgenome=BSgenome.Mmusculus.UCSC.mm10,
                                     txObject=txdb_mouse,
                                     colname="alignments_mouse",
                                     n_mismatches=1)
results_human
```

Our results are stored in the `alignments_mouse` column. We can access these alignments with the `alignments` function and by specifying the `columnName`:

```{r}
alignments(results_human, columnName="alignments_mouse")
```

With these data we can filter our gRNAs for those that target both homologs (and we have off-target annotation for the mouse genome!).

```{r}
aln <- alignments(results_human, columnName="alignments_mouse")
cds_targets <- aln$cds
aln <- aln[!is.na(cds_targets) & cds_targets == "Kras"]
targets_Kras <- unique(names(aln))
results_human <- results_human[targets_Kras]
```

Adding alignments for the human genome (or any other genome) will overwrite the summary columns in `results_human` (`n0`, `n0_c`, `n1`, and `n1_c`) unless we set `addSummary=FALSE` in `addSpacerAlignments`. We should also take care to ensure the column name for our alignments annotation remains unique so it will not be overwritten. Here we add alignment annotation for the human genome, but overwrite the mouse alignment summary columns (see the warning message below).

```{r}
# Path of the hg38 bowtie index on my personal laptop:
bowtie_index_human <- "/Users/hoberecl/crisprIndices/bowtie/hg38/hg38"
# bowtie_index_human <- "/Users/fortinj2/crisprIndices/bowtie/hg38/hg38"

results_human <- addSpacerAlignments(results_human,
                                     aligner="bowtie",
                                     aligner_index=bowtie_index_human,
                                     bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                                     txObject=txdb_human,
                                     colname="alignments_human",
                                     n_mismatches=1)
results_human
```



# Session Info

```{r}
sessionInfo()
```



