---
title: "Using crisprDesign to design gRNAs for optical pooled screening (OPS)"
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin, Luke Hoberecht

Date: `r format(Sys.time(), '%d %B, %Y')`


# Introduction

Optical pooled screening (OPS) combines image-based sequencing (in 
situ sequencing) of gRNAs and optical phenotyping on the same physical 
wells [@ops]. In such experiments, guide RNA (gRNA) spacer sequences 
are partially sequenced from the 5-prime end; the length of these 
truncated sequences, or barcodes, which corresponds to the number of 
sequencing cycles, is fixed and chosen by the experimentalist. From a 
gRNA design perspective, additional constraints are needed to ensure 
sufficient dissimilarity between the rruncated barcodes for their 
identification during the analysis.

This tutorial will demonstrate how to design gRNAs for use in 
optical pooled screens, with emphasis on the constraints described 
above. Common gRNA design steps that are not specific to OPS are omitted 
in this tutorial (e.g. off-target search, or on-target activity prediction)
here. Users can peruse through the list of [available tutorials](https://github.com/crisprVerse/Tutorials) for more information
regarind application-specific gRNA design rules.


# Installation

The Bioconductor packages needed in this tutorial can be installed 
using the following commands:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("crisprDesign")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
```

The GitHub packages needed in this tutorial can be installed using the 
following commands:

```{r, eval=FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")

devtools::install_github("crisprVerse/crisprDesignData")
```


# Terminology

See the [CRISPRko design vignette](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar with the terminology used throughout this tutorial.


# Design for optical pooled screening (OPS)

To illustrate the functionalities of `crisprDesign` for designing 
OPS libraries, we will design a small CRISPRko OPS library targeting 
3 genes of the human RAS family: KRAS, HRAS, and NRAS. We will use the SpCas9
nuclease. 

We will design gRNAs for an experiment that uses 8 in situ sequencing cycles:

```{r}
n_cycles=8
```

## Loading packages

Before we start, we first load the necessary packages for this tutorial:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprBase)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
```


## Creating the GuideSet

We begin by loading the SpCas9 `CrisprNuclease` object from the `crisprBase`
package

```{r}
data(SpCas9, package="crisprBase")
```

as well as data containing gene regions for the human genome:

```{r}
data(txdb_human, package="crisprDesignData")
```

For more information on `txdb_human` and how to create similar gene annotation objects, see the [Building a gene annotation object](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation) tutorial.


Next, we find the CDS coordinates for our genes using the `queryTxObject` function:

```{r}
target_genes <- c("KRAS", "HRAS", "NRAS")
target_regions <- queryTxObject(txdb_human,
                                featureType="cds",
                                queryColumn="gene_symbol",
                                queryValue=target_genes)
```

then build our `GuideSet` with the `findSpacers` function:

```{r}
gs <- findSpacers(target_regions,
                  crisprNuclease=SpCas9,
                  bsgenome=BSgenome.Hsapiens.UCSC.hg38)
```

As we will want to distinguish which gene each spacer targets, we will add `gene_symbol` and `gene_id` columns from `target_regions`.

```{r}
gene_info <- target_regions[gs$region]
gs$gene_symbol <- gene_info$gene_symbol
gs$gene_id <- gene_info$gene_id
```


## Adding OPS barcodes

We can add our OPS barcodes to the GuideSet with the `addOpsBarcodes` function. This function extracts the `n_cycles` nucleotides from the 5-prime end of our spacers and stores them in the `opsBarcode` column:

```{r}
gs <- addOpsBarcodes(gs,
                     n_cycles=n_cycles)
head(gs$opsBarcode)
```


## Barcode distance matrix

We can pass our barcodes to the function `getBarcodeDistanceMatrix` to 
calculate the nucleotide distance between them. The `dist_method` argument
determines the type of distance to calculate: `"hamming"`, which only 
considers substitutions (default) or `"levenstein"`, which also allows for
insertions and deletions. 

As a brief demonstration, let's look at the distances between the first few
barcodes in our `GuideSet`. We set the `binarize` argument (more on this
parameter later) to `FALSE` to show distances:

```{r}
barcodes <- gs$opsBarcode
dist <- getBarcodeDistanceMatrix(barcodes[1:5],
                                 binnarize=FALSE)
dist
```

Note that the output is a sparse matrix, so the barcodes along the diagonal
(i.e., compared against themselves) return `.`, or a distance of zero. 
To compare one set of barcodes against another, we can pass the other set
to the `targetBarcodes` argument (the former barcode set being passed to 
the `queryBarcodes` argument, which is compared against itself when `targetBarcodes` is `NULL`):

```{r}
dist <- getBarcodeDistanceMatrix(barcodes[1:5],
                                 targetBarcodes=barcodes[6:10],
                                 binnarize=FALSE)
dist
```

The question we are interested in with respect to barcode distances is 
whether this distance is sufficiently dissimilar for accurate identification 
of spacers during sequencing. This minimum distance edit (`min_dist_edit`) 
relies on the accuracy of various steps in the experiment. Suppose,
as a conservative estimate, that we can expect no more than two edits 
per barcode in our example. A `min_dist_edit` of `3` should 
suffice. Setting the `binnarize` argument to `TRUE`, and passing our minimum distance edit value to `min_dist_edit` will binarize the output, flagging barcodes (with a value of `1`) that are too similar and should not both be included in our library:

```{r}
dist <- getBarcodeDistanceMatrix(barcodes[1:5],
                                 barcodes[6:10],
                                 binnarize=TRUE,
                                 min_dist_edit=3)
dist
```

Using this function with large sets of barcodes can be taxing on memory.
To manage this, it is recommended to set `splitByChunks=TRUE` and specify 
the number of chunks with `n_chunks` (see `?getBarcodeDistanceMatrix`).


## Designing OPS libraries

The `designOpsLibrary` function allows users to perform a complete 
end-to-end OPS library design. We will design our library with 4 
gRNAs per gene using the `n_guides` and `gene_field` (to identify gRNAs
by gene target) parameters. We will also use the same distance method 
and minimum distance edit parameters as in the example above.

NOTE: it is advised to first complete other steps in gRNA design
(annotating, filtering, and ranking gRNAs in the `GuideSet`) prior to 
using this function; this will ensure the library contains the best 
gRNAs. As this example did not rank gRNAs, we are notified that 
rankings are assigned by the order in which gRNAs appear in our input.

```{r}
df <- data.frame(ID=names(gs),
                 spacer=gs$protospacer,
                 opsBarcode=gs$opsBarcode,
                 gene_symbol=gs$gene_symbol)
opsLibrary <- designOpsLibrary(df,
                               n_guides=4,
                               gene_field="gene_symbol",
                               min_dist_edit=5,
                               dist_method="hamming")
opsLibrary
```


## Adding gRNAs to an existing OPS library

Suppose we later wish to add another gene target to our library, 
but also want to retain the gRNAs that are currently in our library.
We can append these additional gRNAs with the `updateOpsLibrary` function. 
This function has the same parameters as `designOpsLibrary`, with an 
additional `opsLibrary` argument to which we pass our original OPS library.

To demonstrate, we will add the MRAS gene to our library. We first
construct the `GuideSet` for MRAS:

```{r}
target_region <- queryTxObject(txdb_human,
                               featureType="cds",
                               queryColumn="gene_symbol",
                               queryValue="MRAS")
gs_mras <- findSpacers(target_region,
                       crisprNuclease=SpCas9,
                       bsgenome=BSgenome.Hsapiens.UCSC.hg38)
gs_mras$gene_symbol <- "MRAS"
gs_mras$gene_id <- "ENSG00000158186"
```

then add barcodes and construct the `data.frame`:

```{r}
## add OPS barcodes
gs_mras <- addOpsBarcodes(gs_mras,
                          n_cycles=n_cycles)

## construct data.frame
df_mras <- data.frame(ID=names(gs_mras),
                      spacer=gs_mras$protospacer,
                      opsBarcode=gs_mras$opsBarcode,
                      gene_symbol=gs_mras$gene_symbol)
```

which we then pass with our other parameters to `updateOpsLibrary`:

```{r}
opsLibrary <- updateOpsLibrary(opsLibrary,
                               df_mras,
                               n_guides=4,
                               gene_field="gene_symbol",
                               min_dist_edit=5,
                               dist_method="hamming")
opsLibrary
```



# Session Info

```{r}
sessionInfo()
```


# References
