---
title: "Building genome indices for short read aligners"
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin, Luke Hoberecht

Date: `r format(Sys.time(), '%d %B, %Y')`


# Introduction

This vignette demonstrates how to build genome indices for use with the short read aligners bowtie [@langmead2009bowtie], as used by the `Rbowtie` and `crisprBowtie` packages, and BWA-backtrack [@bwa], as used by the `Rbwa` and `crisprBwa` packages.


# Installation

### OS Requirements

The `crisprBowtie` package is supported for macOS, Linux and Windows machines.
The `crisprBwa` package is supported for macOS and Linux only.
Both packages were developed and tested on R version 4.2.

### R Dependencies 

- RBowtie: https://bioconductor.org/packages/release/bioc/html/Rbowtie.html
- RBwa: https://github.com/Jfortin1/Rbwa


# Building a genome index

A genome index file is necessary to use the aligner functions in the `crisprBowtie` package (`runBowtie` and `runCrisprBowtie`) and the `crisprBwa` package (`runBwa` and `runCrisprBwa`). For a given genome, this step only needs to be done once.


## Bowtie index

The `Rbowtie` package offers the convenient function `bowtie_build`, which builds a bowtie index for any custom genome from a FASTA file. In the following example, we build a bowtie index for a small region of the human chromosome 1 (provided in the `crisprBowtie` package) and save the index file to a temporary directory:

```{r}
library(Rbowtie)
fasta <- file.path(find.package("crisprBowtie"), "example/chr1.fa")
tempDir <- tempdir()
Rbowtie::bowtie_build(fasta,
                      outdir=tempDir,
                      force=TRUE,
                      prefix="myIndex")
```

See the [crisprBowtie](https://github.com/Jfortin1/crisprBowtie) package for information on how to obtain alignments using this index file.


## Bwa index

Building a BWA index is made simple with the `bwa_build_index` function from the `Rbwa` package. This function builds the index file for any custom genome from a FASTA file. As an example, we build a BWA index for a small portion of the human chromosome 12 (provided in the `crisprBwa` package) and save the index file to a temporary directory:

```{r}
library(Rbwa)
fasta <- system.file(package="crisprBwa", "example/chr12.fa")
outdir <- tempdir()
index <- file.path(outdir, "chr12")
Rbwa::bwa_build_index(fasta,
                      index_prefix=index)
```

See the [crisprBwa](https://github.com/Jfortin1/crisprBwa) package for information on how to obtain alignments using this index file.


# Building a transcriptome index

For applications using RNA-targeting nucleases such as CasRx, it is preferable to search for alignments against the transcriptome rather than the entire genome. To build an index file for the transcriptome we must first generate a FASTA file containing the transcriptome. This is easily accomplished with the `crisprDesign` function `getMrnaSequences`, as shown in the following example:

```{r, eval=FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)
library(crisprDesign)
library(crisprDesignData)
data("txdb_human", package="crisprDesignData")
exon_ids <- unique(txdb_human$exons$tx_id)
mrnasHuman <- getMrnaSequences(exon_ids,
                               bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                               txObject=txdb_human)
library(Biostrings)
writeXStringSet(mrnasHuman,
                file="ensembl_human_104.fasta",
                format="fasta")
```

Note that the `seqnames` of this FASTA file are Ensembl transcript IDs instead of chromosomes. Once the FASTA file has been generated, the process for constructing either a bowtie or BWA index file is the same as described in the above sections.


# Reproducibility

```{r}
sessionInfo()
```



# References

