---
title: "Building a genome-wide gRNA database"
author: Jean-Philippe Fortin, Luke Hoberecht
output: 
  github_document:
    toc: true
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```


# Introduction

In this tutorial, we provide reproducible code to design and annotate
gRNAs against all human protein-coding genes using the nuclease SpCas9. 

# Loading necessary packages

We first load the necessary packages:

```{r, warning=FALSE, message=FALSE}
library(crisprBase)
library(crisprScore)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
```


### Specifying the genome

We specify a `BSGenome` object that contains the DNA sequence of the human
genome in hg38 coordinates:

```{r}
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
```

### Specifying the genome index

We specify the file path of the Bowtie index that we will need for
off-target alignment:

```{r}
bowtie_index <- "/Users/fortinj2/crisprIndices/bowtie/hg38/hg38"
```

For instructions on how to build a Bowtie index from a given reference
genome, see the [genome index tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Genome_Indices). 



### Specifying a SNP VCF file

To flag gRNAs overlapping common SNPs, we specify a VCF file obtained from the
dbSNP website containing common SNPs from the dbSNP151 release:

```{r}
vcf <- "/Users/fortinj2/crisprIndices/snps/dbsnp151.grch38/00-common_all.vcf.gz"
```

The VCF file was obtained from [NCBI](https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf).

### Specifying the nuclease

We load a `CrisprNuclease` object representing the SpCas9 nuclease from 
the `crisprBase` package:

```{r}
data(SpCas9, package="crisprBase")
crisprNuclease <- SpCas9
```

To learn how to specify or build a custom nuclease, see the [nuclease tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Custom_Nuclease).



### Specifying on-target scoring methods

We specify which on-target scoring methods should be used to score 
the gRNAs:

```{r}
scoring_methods <- c("deephf", "deepspcas9")
```


One can see which scoring methods are available for a given nuclease using
the following command:


```{r}
crisprScore::scoringMethodsInfo
```


### Specifying gene models and TSS annotations

To annotate gRNAs with a gene and TSS annotation, we need to specify a gene model 
formatted as a `GRangesList` object, as well as a TSS annotation with a 
`GRanges` object. The `crisprDesignData` contains such objects for both 
the human and mouse genomes, in GRCh38 (hg38) and GRCm38 (mm10) coordinates,
respectively. Ensembl gene models were used to generate such objects.
We load those objects:


```{r, warning=FALSE, message=FALSE}
data(txdb_human, package="crisprDesignData")
data(tss_human, package="crisprDesignData")
txObject <- txdb_human
tssObject <- tss_human
```

See the [gene annotation tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation) 
to learn how to build such objects. The [crisprDesignData](https://github.com/crisprVerse/crisprDesignData) also has
tons of useful information. 


### Specifying repeat elements

To avoid designing gRNAs targeting repeat elements, we will specify a `GRanges`
object containing repeats coordinates for the human genome. Here, we use the object `gr.repeats.hg38` in `crisprDesignData`. It contains genomic coordinates of 
the RepeatMasker UCSC track, for the hg38 reference genome:

```{r, warning=FALSE, message=FALSE}
data(gr.repeats.hg38, package="crisprDesignData")
grRepeats <- gr.repeats.hg38
```


# Building a complete annotation for a given gene 


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# Explicitely calling those packages here instead of
# printout when calling precomputeGuides. This can be omitted. 
library(crisprScoreData)
library(ExperimentHub)
library(AnnotationHub)
library(BiocFileCache)
library(dbplyr)
```

The `designCompleteAnnotation` function in `crisprDesign` provides a 
one-step workflow to design and annotate all gRNAs targeting a given gene.
The function was designed to be as comprehensive as possible to 
design and annotate gRNAs in one step. It does the following: 

- Extract the DNA/RNA sequences with `queryTss`/`queryTxDB`
- Design gRNAs with `findSpacers`
- Remove gRNAs targeting repeat elements with `removeRepeats`
- Characterize spacer sequences with `addSequenceFeatures`
- Find on- and off-targets with `addSpacerAlignmentsIterative`
- Add gene annotation with `addGeneAnnotation`
- Add TSS annotation with `addTssAnnotation`
- Add on-target efficiency scores with `addOnTargetScores`
- Add off-target specificity scores with `addOffTargetScores`
- Add SNP annotation with `addSNPAnnotation`
- Add restriction enzymes information with `addRestrictionEnzymes`


Here, we design all CRISPRko gRNAs targeting the 
human KRAS gene (ENSG00000133703):



```{r, warning=FALSE}
gs <- designCompleteAnnotation(queryValue="ENSG00000133703",
                               queryColumn="gene_id",
                               modality="CRISPRko",
                               bsgenome=bsgenome,
                               bowtie_index=bowtie_index,
                               crisprNuclease=SpCas9,
                               txObject=txObject,
                               tssObject=tssObject,
                               grRepeats=grRepeats,
                               vcf=vcf,
                               n_mismatches=1,
                               scoring_methods=scoring_methods)
```


The resulting object is a `GuideSet` object. To learn more about what are
`GuideSet` objects, and how to interact with them, see the [CRISPRko gRNA design tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9).


```{r}
gs
```


### Converting the `GuideSet` object to a list of data.frames

The `flattenGuideSet` function in `crisprDesign` is a convenience function 
to convert a `GuideSet` object into a set of `data.frames` that can be 
saved as plain text files:

```{r}
dfs <- flattenGuideSet(gs)
```

We can look at the names of the data.frames:

```{r}
names(dfs)
```

As an example, let's look at the first rows of the primary data.frame:

```{r}
head(dfs$primary)
```

# Building a complete gRNA database across all protein-coding genes

We first get all possibles genes from our gene model:

```{r}
gene_ids <- unique(txObject$cds$gene_id)
head(gene_ids)
```


and specify where to save the `GuideSet` objects:


```{r, eval=FALSE}
dir <- "./crisprko_cas9_hg38"
if (!dir.exists(dir)){
    dir.create(dir, recursive=TRUE)
}
```


We are now looping over all genes to generate the data:

```{r, eval=FALSE}
lapply(gene_index, function(gene){
    gs <- designCompleteAnnotation(queryValue=gene,
                                   queryColumn="gene_id",
                                   modality="CRISPRko",
                                   bsgenome=bsgenome,
                                   bowtie_index=bowtie_index,
                                   crisprNuclease=SpCas9,
                                   txObject=txObject,
                                   tssObject=tssObject,
                                   grRepeats=grRepeats,
                                   vcf=vcf,
                                   n_mismatches=3,
                                   scoring_methods=scoring_methods)
    write.rds(gs, file=file.path(dir, paste0(gene, ".rds")))
})
```

This loop can be modified by the user to use an embarrassingly-parallel 
approach, using the [BiocParallel](https://bioconductor.org/packages/BiocParallel/) package, for instance.


Building a database for CRISPRa and CRISPRi applications works similarly
See `?designCompleteAnnotation` for more information. 


# Reproducibility

```{r}
sessionInfo()
```
