---
title: "Building a gene annotation object"
author: Jean-Philippe Fortin, Luke Hoberecht
output: 
  github_document:
    toc: true
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```



# Introduction

In this tutorial, we describe the process for making and using  
rich gene annotation objects to be used throughout the crisprVerse ecosystem. 
Such objects enable users to retrieve coordinates of transcripts, exons, etc.
Those objects are also used by several functions in the [crisprDesign package](https://github.com/crisprVerse/crisprDesign) to add gene annotations
to both gRNA on-targets and off-targets. This is what the `txObject` argument 
in many of the functions expect. 

We will also describe the process for constructing and using a 
transcription start site (TSS) annotation object (`tssObject` argument in many
of the functions).  


# Installation

See the [Installation tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Installation) to learn how to install the packages  `crisprDesign` and `crisprDesignData` required in this tutorial. 


### Getting started

The packages can be loaded into an R session in the usual way:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprDesign)
library(crisprDesignData)
```

# Building gene annotation objects

In the crisprVerse, we represent gene annotations using `GRangesList` object,
and this can be easily constructed using the commonly-used Bioconductor
objects `TxDb` (see the [GenomicFeatures package](https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf) to learn more about `TxDb` objects).
We will now show several ways of constructing such objects. 


# Building a GRangesList from Ensembl

We construct a gene annotation object for the human genome
using the Ensembl release 104 (hg38). This can be done using the function
`getTxDb` in `crisprDesign`:

```{r, warning=FALSE, message=FALSE, results='hide'}
txdb <- getTxDb(organism="Homo sapiens", release=104)
```

This may take several minutes, and note that this requires an internet connection.
In case it times out, one can increase the timeout option using the following:

```{r}
options(timeout = max(10000000, getOption("timeout")))
```

Once obtained, we can convert the object
into a `GRangesList` using the function `TxDb2GRangesList` from `crisprDesign`:

```{r, warning=FALSE, message=FALSE, results='hide'}
grList <- TxDb2GRangesList(txdb)
```

We will specify that the genome is hg38:

```{r}
GenomeInfoDb::genome(grList) <- "hg38"
```

And that's it! The `grList` object contains all of the information about
the Ensembl release 104 gene model, and is ready to be used in the crisprVerse. 
Let's take a quick look at our gene annotation object:

```{r}
names(grList)
grList$transcripts
```



## Building a tssObject

Building a TSS annotation object requires only one additional step after
constructing the `GRangesList` object described above.
This can be obtained using the function `getTssObjectFromTxObject` in
`crisprDesign`:

```{r}
tssObject <- getTssObjectFromTxObject(grList)
tssObject
```


# Using gene annotation objects

The gene (or TSS) annotation objects described above are often necessary for 
the full characterization of CRISPR gRNAs as they as inputs for several
of the `crisprDesign` functions, including `queryTxObject`,
`queryTssObject`, `addGeneAnnotation`, `addTssAnnotation`, 
and `addSpacerAlignments`. 

For convenience, we provide in the [crisprDesignData package](https://github.com/crisprVerse/crisprDesignData)
precomputed gene annotation for human and mouse:

|  Object name | Object class | Version | Description |
|----------- | ----------- | ----------- |----------- | 
| `txdb_human`  | `GRangesList`   | Release 104 | Ensembl gene model for human (hg38/GRCh38) |
| `txdb_mouse`  | `GRangesList`   | Release 102 | Ensembl gene model for mouse (mm10/GRCm38) |
| `tss_human`   | `GRanges`       | Release 104 | Ensembl-based TSS coordinates for human (hg38/GRCh38) |
| `tss_mouse`   | `GRanges`       | Release 102 | Ensembl-based TSS coordinates for human (mm10/GRCm38) |

# Building a gene annotation object from a GFF file

If you have a General Feature Format (GFF) file from which you want to 
construct the gene annotation object, you can pass this to the `file` argument 
of the `crisprDesign` function `getTxDb`; this will create the `TxDb` object 
using the `GenomicFeatures` function `makeTxDbFromGFF`.


# Reproducibility

```{r}
sessionInfo()
```


