---
title: "Using crisprDesign to design paired gRNAs"
author: Jean-Philippe Fortin, Luke Hoberecht
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```


# Introduction

In this tutorial, we illustrate the main functionalities of  `crisprDesign` 
for designing pairs of gRNAs. 

# Getting started

# Installation

See the [Installation tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Installation) to learn how to install the packages necessary for this tutorial:
`crisprDesign`, `crisprDesignData`


## Terminology

See the [CRISPRko Cas9 design tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar with the terminology used throughout this tutorial.

## Paired gRNA design overview

There are several applications that require the design of gRNA pairs:

1. Double nicking with CRISPR/Cas9 [@ran2013double]
2. Dual-promoter screening systems [@han2017synergistic]
3. Multiplexing gRNAs with enAsCas12a [@deweirdt2021optimization]
4. Nanopore Cas9-targeted sequencing (nCATS) [@gilpatrick2020targeted]

The `crisprDesign` package provides an infrastructure to store an annotate
gRNA pairs via the  `PaireGuideSet` object, which behaves very similarly
to the `GuideSet` object used for unpaired gRNAs. We designed the 
functionalities for paired gRNAs with the aforementioned applications in mind. 

In this tutorial, we will go through a simple example to illustrate 
the general concept behind paired gRNA design with `crisprDesign`. 


# A simple example: deleting a KRAS exon with a pair of gRNAs

We will show here how to design an optimal pair of Cas9 gRNAs flanking the
second exon of the human gene KRAS (ENSG00000133703), with the goal of 
creating a deletion that will excise the exon. 

We first start by loading the necessary packages:

```{r, message=FALSE, warning=FALSE, results='hide' }
library(crisprDesign)
library(crisprDesignData)
library(crisprBase)
library(BSgenome.Hsapiens.UCSC.hg38)
```

We will be designing gRNAs for the SpCas9 nuclease, which can be loaded 
from We load the `crisprBase` package (see the `crisprBase` [vignette](https://github.com/crisprVerse/crisprBase) for 
instructions on how to create or load alternative nucleases):


```{r}
data(SpCas9, package="crisprBase")
```

Let's get the genomic coordinates of the second exon.
First, we obtain from `crisprDesignData` a `GRangesList` object that defines
the genomic coordinates (hg38 genome) of human protein-coding genes:

```{r}
data(txdb_human, package="crisprDesignData")
```

We then get the exonic coordinates of the canonical transcript ENST00000311936 
using the function `queryTxObject` from `crisprDesign`:


```{r}
exons <- queryTxObject(txObject=txdb_human,
                       featureType="exons",
                       queryColumn="tx_id",
                       queryValue="ENST00000311936")
exons
```

Finally, we select the second exon:

```{r}
exon <- exons[exons$exon_rank==2]
names(exon) <- "exon_kras"
exon
```

The exon is on chr12, and spans the region 25245274-25245395 (122 
nucleotides in length). We aim to design gRNAs pairs for which one 
gRNA is located upstream of the exon, and another located downstream 
of the exon. To be able to find good gRNA candddates, let's define those 
regions to have 100 nucleotides on each side:


```{r}
library(IRanges)
regionUpstream   <- IRanges::flank(exon, width=100, start=FALSE)
regionDownstream <- IRanges::flank(exon, width=100, start=TRUE)
names(regionUpstream) <- "upstreamTarget"
names(regionDownstream) <- "downstreamTarget"
```

Similar to the `findSpacers` function in `crisprDesign`, we will need
to specify a `BSgenome` object containing the reference genome DNA
sequences:

```{r}
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
```

We are now ready to find all candidate gRNA pairs:

```{r}
pairs <- findSpacerPairs(x1=regionUpstream,
                         x2=regionDownstream,
                         bsgenome=bsgenome,
                         crisprNuclease=SpCas9)
```

The `x1` and `x2` arguments specify the genomic regions in which gRNAs at
position 1 and position 2 should be targeting, respectively. 
The function finds all possible pair combinations between spacers 
found in the region specified by `x1` and spacers found in the region s
pecified by `x2`. Let' first name our pairs:

```{r}
names(pairs) <- paste0("pair_", seq_along(pairs))
```

Let's see what the results look like:
```{r}
head(pairs, n=3)
```

The returned object is a `PairedGuideSet`, which can be though of a list 
of two `GuideSet` objects. The first and second `GuideSet` store 
information about gRNAs at position 1 and position 2, respectively. 
They can be accessed using the `first` and `second` functions:

```{r}
grnas1 <- first(pairs)
head(grnas1, n=3)
```

and 

```{r}
grnas2 <- second(pairs)
head(grnas2, n=3)
```

The `pamOrientation` function returns the PAM orientation of the pairs:

```{r}
head(pamOrientation(pairs))
```

and takes 4 different values: `in` (for PAM-in configuration), `out` (for 
PAM-out configuration), `fwd` (both gRNAs target the forward strand), 
and `rev` (both gRNAs target the reverse strand); see figure below for 
an illustration of the PAM orientations for the SpCas9 nuclease. 
The importance of the PAM orientation is application-specific. For 
Nanopore Cas9-targeted sequencing, PAM-in configuration is preferred.
For double nicking with CRISPR/Cas9, PAM-out configuration is preferred.
For applications using a dual-promoter system, no configuration is 
preferred. 


```{r, echo=FALSE, out.width = "75%", fig.align="center", fig.cap="Different PAM orientations for Cas9 paired gRNAs"}
knitr::include_graphics("./figures/paired_simplified.svg")
```


The function `pamDistance` returns the distance between the PAM sites of the
two gRNAs. The function `cutLength` returns the distance between the 
cut sites of the two gRNAs, and the function `spacerDistance` 
returns the distance between the two spacer sequences of the gRNAs.

Most functionalities available for designing single gRNAs (`GuideSet` 
annotation functions described in [this tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9)) work similarly for `PairedGuideSet` objects. This includes:

- `addSequenceFeatures`
- `addSpacerAlignments`
- `addGeneAnnotation`
- `addTssAnnotation`
- `addOnTargetScores`
- `addOffTargetScores`
- `addPamScores`
- `addSNPAnnotation`
- `addRestrictionEnzymes`
- `addCompositeScores`
- `addConservationScores`

Each function adds an annotation to the first and second `GuideSet` objects
stored in the `PairedGuideSet`. Let's look at an example using 
`addSequenceFeatures`:

```{r}
pairs <- addSequenceFeatures(pairs)
```

and let's look at the `GuideSet` in the first position:

```{r}
head(first(pairs), n=3)
```

This comes in handy to filter out pairs with unwanted sgRNA characteristics, 
e.g. sgRNA with polyT stretches:

```{r}
good1 <- !first(pairs)$polyT
good2 <- !second(pairs)$polyT
pairs <- pairs[good1 & good2]
```

To select the final candidate pairs to excise the KRAS exon, we will
filter out pairs with low  predicted on-target activity using the
DeepHF on-target activity score. We first add the score:

```{r}
pairs <- addOnTargetScores(pairs, methods="deephf")
```

and only keep pairs for which both gRNAs have a score greater than 0.5:

```{r}
good1 <- first(pairs)$score_deephf>=0.5
good2 <- second(pairs)$score_deephf>=0.5
pairs <- pairs[good1 & good2]
```

This leaves us with 2 candidate pairs:

```{r}
pairs
```


Finally, let's check for off-targets. 
We need to specify the path of the bowtie index that was generated 
from the human reference genome:

```{r}
bowtie_index <- "/Users/fortinj2/crisprIndices/bowtie/hg38/hg38"
```

For instructions on how to build a Bowtie index from a given reference genome, see the [genome index tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Genome_Indices) or the [crisprBowtie page](https://github.com/crisprVerse/crisprBowtie). 

We are now ready to search for off-targets with up to 3 mismatches:


```{r}
pairs <- addSpacerAlignments(pairs,
                             txObject=txdb_human,
                             aligner_index=bowtie_index,
                             bsgenome=bsgenome,
                             n_mismatches=3)
```

We are in luck, none of the spacer sequences has an off-target in the coding 
region of other genes:


```{r}
good1 <- first(pairs)$n1_c==0 & first(pairs)$n2_c==0 & first(pairs)$n3_c==0
good2 <- second(pairs)$n1_c==0 & second(pairs)$n2_c==0 & second(pairs)$n3_c==0
pairs <- pairs[good1 & good2]
pairs
```


One can get the spacer sequences using the `spacers` accessor function
as usual:

```{r}
spacers(pairs)
```



# Session Info

```{r}
sessionInfo()
```

# References


