---
title: "Using crisprDesign to design gRNAs for CRISPRkd with CasRx"
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```

Authors: Jean-Philippe Fortin, Luke Hoberecht

Date: `r format(Sys.time(), '%d %B, %Y')`


# Introduction

`crisprDesign` is a comprehensive software package for designing and annotating CRISPR guide RNA (gRNA) sequences, including the characterization of on-targets and off-targets, gene context annotation, and SNP annotation (human only). The software was developed to be as applicable and generalizable as possible. It currently support four types of CRISPR modalities (modes of perturbations): CRISPR knockout (CRISPRko), CRISPR activation (CRISPRa), CRISPR inhibition (CRISPRi) and CRISPR base editing (CRISPRbe) (see @crispracrisprireview for a review of CRISPR modalities).

This package utilizes the `crisprBase` package to enable gRNA design for any CRISPR nuclease via the `CrisprNuclease` class. Nucleases that are commonly used in the field are provided, including DNA-targeting nucleases (e.g. SpCas9, AsCas12a) and RNA-targeting nuclease (e.g. CasRx (RfxCas13d)).

`crisprDesign` is fully developed to work with the genome of any organism, and can also be used to design gRNAs targeting custom DNA sequences.


# Installation


The Bioconductor packages needed in this vignette can be downloaded using
the following commands:


```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("crisprDesign")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg38")
```

The GitHub packages needed in this vignette can be downloaded using
the following commands:

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("Jfortin1/crisprDesignData")
```


Users interested in contributing to `crisprDesign` might want to look at the following CRISPR-related package dependencies:

- `crisprBase`: core CRISPR functions and S4 objects
- `crisprBowtie`: aligns gRNA spacers to genomes using the ungapped 
aligner `bowtie`
- `crisprBwa`: aligns gRNA spacers to genomes using the ungapped 
aligner `BWA`
- `crisprScore`: implements state-of-the-art on- and off-target scoring 
algorithms
- `crisprScoreData`: pre-trained models necessary for `crisprScore`

You can contribute to the package by submitting pull requests to our [GitHub repo](https://github.com/Jfortin1/crisprDesign). 


# Terminology

See the [CRISPRko design vignette](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar with the terminology used throughout
this tutorial.


# CRISPR knockdown with Cas13d

## Loading packages

The following examples use the `crisprBase`, `crisprDesign`, `crisprDesignData`, and `BSgenome.Hsapiens.UCSC.hg38` packages. Before we begin, let's load the 
necessary packages:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprBase)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
```


## Creating the GuideSet

In this example, we will design gRNAs for an RNA-targeting nuclease using `crisprDesign`, specifically, we will use the CasRx (Cas13d) nuclease [@cas13d] to target the primary isoform of the human KRAS gene. In contrast to DNA-targeting nucleases, the target spacers are composed of mRNA sequences rather than DNA genomic sequences. [disclosure...]

We begin by loading the CasRx `CrisprNuclease` object from `crisprBase`:

```{r}
data(CasRx, package="crisprBase")
CasRx
```

The PFS sequence (the equivalent of a PAM sequence for RNA-targeting nucleases) for CasRx is `N`, meaning there is no specific PFS sequences preferred by CasRx. 

Next, we extract the mRNA sequence for our target transcript (Ensembl transcript ID: ENST00000311936). For this we need the coordinates for our transcript, which are conveniently stored in the `crisprDesignData` package as `txdb_human` (for more information on `txdb_human` and how to create similar gene annotation objects, see the [Building a gene annotation object](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation) tutorial), as well as a `BSgenome` object containing the exonic sequences.

```{r}
data("txdb_human", package="crisprDesignData")
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
```

We obtain our mRNA sequence with `getMrnaSequences`:

```{r}
txid <- "ENST00000311936"
mrna <- getMrnaSequences(txids=txid,
                         bsgenome=bsgenome,
                         txObject=txdb_human)
mrna
```

We then use the function `findSpacers` to design our gRNAs. For the sake of brevity we will only consider a subset of 100 gRNAs:

```{r}
gs <- findSpacers(mrna[["ENST00000311936"]],
                  crisprNuclease=CasRx)
gs <- gs[1000:1100]
head(gs)
```

Note that all protospacer sequences are located on the original strand of the mRNA sequence. For RNA-targeting nucleases, the spacer and protospacer sequences are the reverse complement of each other. (Compare the output of the code below with a `GuideSet` that uses a DNA-targeting nuclease--for such `GuideSet`s, the output of `spacers` and `protospacers` are identical.)

```{r}
head(spacers(gs))
head(protospacers(gs))
```


## Annotating the GuideSet

Next, we annotate our candidate gRNAs to assess quality. There are several 
functions in `crisprDesign` that provide annotation for features that are 
nonspecific to CRISPRkd, for which we refer the reader to the
[CRISPRko design with Cas9](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) tutorial for more information. The sections below will 
cover annotation functions that are of particular interest to,
or deserve extra care for CRISPRkd applications.

### Adding spacer alignments

Since our CRISPR nuclease targets RNA rather than DNA, off-target searches should be restricted to the transcriptome. We can perform such a search using one of two methods. For the first method, we set the `aligner` argument to `"biostrings"` and pass a `DNAStringSet` representation of the transcriptome to the argument `custom_seq`. We can create this representation with `getMrnaSequences` and all transcript IDs found in `txdb_human`. For the sake brevity and demonstration, however, we will limit our search to mRNAs belonging to the KRAS gene. We will also search for off-targets having up to one mismatch and pass `txdb_human` to the `txObject` argument so that our alignments will be accompanied with gene annotation. 

```{r, warning=FALSE, message=FALSE, results='hide'}
tx_ids <- c("ENST00000256078", "ENST00000311936",
            "ENST00000557334", "ENST00000556131")
kras_mrnas <- getMrnaSequences(txids=tx_ids,
                               bsgenome=bsgenome,
                               txObject=txdb_human)
gs <- addSpacerAlignments(gs,
                          aligner="biostrings",
                          txObject=txdb_human,
                          n_mismatches=1,
                          custom_seq=kras_mrnas)
```
```{r}
tail(gs)
```

The columns `n0_gene` and `n0_tx` report the number of on-targets at the gene- and transcript-level, respectively. For instance, for all spacers shown above `n0_tx` is equal to 3, meaning they all map to three isoforms of KRAS (since our `custom_seq` is made up of KRAS isoforms only). We can use the `onTargets` accessor function to look at each alignment for the first spacer:

```{r}
# onTargets(gs["spacer_1095"]) # bug: incorrect spacer alignments selected
onTargets(gs)[names(onTargets(gs)) == "spacer_1095"]
```

The second method uses the `bowtie` (or `bwa`) aligner. This requires building a transcriptome bowtie (or BWA) index file first. See the [Building genome indices for short read aligners](https://github.com/crisprVerse/Tutorials/tree/master/Building_Genome_Indices) tutorial for more information. 


# Session Info

```{r}
sessionInfo()
```


# References

