---
title: "Using crisprDesign to design gRNAs for CRISPRbe"
author: Jean-Philippe Fortin, Luke Hoberecht
output: 
  github_document:
    toc: true
bibliography: references.bib
---

```{r, echo=FALSE, results="hide"}
options("knitr.graphics.auto_pdf"=TRUE)
```


# Introduction

In this tutorial, We illustrate the CRISPR base editing (CRISPRbe) functionalities 
of `crisprDesign` by designing and characterizing gRNAs targeting the 
human gene KRAS using the cytidine base editor BE4max [@koblan2018improving]. 


# Installation

See the [Installation tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Installation) to learn how to install the packages necessary for this tutorial:
`crisprDesign`, `crisprDesignData`


# Terminology

See the [CRISPRko design tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) to get familiar 
with the terminology used throughout this tutorial.


# CRISPR base editing with BE4max

## Loading packages

We first load the necessary packages for this tutorial:

```{r, warning=FALSE, message=FALSE, results='hide'}
library(crisprBase)
library(crisprDesign)
library(crisprDesignData)
library(BSgenome.Hsapiens.UCSC.hg38)
```


## Creating the GuideSet

We first load the BE4max `BaseEditor` object from the `crisprBase` package:

```{r}
data(BE4max, package="crisprBase")
BE4max
```

The editing probabilities of the base editor BE4max are stored in a 
matrix where rows correspond to the different nucleotide substitutions, 
and columns correspond to the genomic coordinate relative to the PAM site. 
The `editingWeights` function from `crisprBase` retrieves those probabilities. 
One can see that C to T editing is optimal around 15 nucleotides upstream of 
the PAM site for the BE4max base editor:

```{r}
crisprBase::editingWeights(BE4max)["C2T",]
```

Let's create the `GuideSet` containing gRNAs targeting KRAS.

We first load the data containing gene regions for the human genome 
from `crisprDesignData`:

```{r}
data("txdb_human", package="crisprDesignData")
```

For more information on `txdb_human` and how to create 
similar gene annotation objects, see the [Building a gene annotation object tutorial](https://github.com/crisprVerse/Tutorials/tree/master/Building_Gene_Annotation).


We will also load the `BSgenome` package containing DNA sequences for the hg38
genome:

```{r, message=FALSE, warning=FALSE, results='hide'}
library(BSgenome.Hsapiens.UCSC.hg38)
```


We retrive the genomic coordinates of the KRAS CDS
```{r}
gr <- queryTxObject(txObject=txdb_human,
                    featureType="cds",
                    queryColumn="gene_symbol",
                    queryValue="KRAS")
```

and design all possigle gRNAs using the function `findSpacers`:

```{r}
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
gs <- findSpacers(gr,
                  bsgenome=bsgenome,
                  crisprNuclease=BE4max)
```


## Annotating the GuideSet

Next, we annotate our candidate gRNAs to assess quality. There are several 
functions in `crisprDesign` that provide annotation for features that are 
nonspecific to CRISPRbe, for which we refer the reader to the
[CRISPRko design with Cas9](https://github.com/crisprVerse/Tutorials/tree/master/Design_CRISPRko_Cas9) tutorial for more information. The sections below will 
cover annotation functions that are of particular interest to,
or deserve extra care for CRISPRbe applications.


### Adding edited alleles

The function `addEditedAlleles` finds, characterizes, and scores 
predicted edited alleles for each gRNA and a chosen transcript. 
It requires a transcript-specific annotation that can be obtained 
with the `getTxInfoDataFrame` function. Here, we perform the analysis
using the primary isoform of KRAS (Ensembl transcript ID: ENST00000311936).

We first get the transcript table for our transcript

```{r}
txid <- "ENST00000311936"
txTable <- getTxInfoDataFrame(tx_id=txid,
                              txObject=txdb_human,
                              bsgenome=bsgenome)
head(txTable)
```

and then add the edited alleles annotation to the `GuideSet`:

```{r}
editingWindow <- c(-20,-8)
gs <- addEditedAlleles(gs,
                       baseEditor=BE4max,
                       txTable=txTable,
                       editingWindow=editingWindow)
```

The `editingWindow` argument specifies the window of editing that we
are interested in. When not provided, it uses the default window provided 
in the `BaseEditor` object. Note that providing large windows can 
exponentially increase computing time as the number of possible 
alleles grows exponentially. 

Let's retrieve the edited alleles for the first gRNA:

```{r}
alleles <- editedAlleles(gs)[[1]]
```

We get a `DataFrame` object with useful metadata:

```{r}
metadata(alleles)
```

The `wildtypeAllele` reports the unedited nucleotide sequence of the
region specified by the editing window (with respect to the gRNA PAM site).
It is always reported from the 5' to 3' direction on the strand corresponding 
to the gRNA strand. The `start` and `end` fields specify the
corresponding coordinates on the transcript. 

Let's look at the edited alleles:

```{r}
head(alleles)
```

The `DataFrame` is ordered by descending values in the `score` column.
This `score` represents the likelihood of the edited allele to occur 
relative to all possible edited alleles, and is calculated using the 
editing weights stored in the `BE4max` object. The `seq` column represents
the edited nucleotide sequences. As with the `wildtypeAllele` in the metadata, 
they are always reported from the 5' to 3' direction on the strand 
corresponding to the gRNA strand. 

The `variant` column describes the
functional consequence of the editing event (silent, nonsense or
missense mutation). If an edited allele results in multiple editing 
events, as can happen when multiple bases are edited, the most 
consequential mutation (nonsense over missense, missense over silent) 
is reported. Finally, the `aa` column reports the resulting edited 
amino acid sequence, with each single letter code mapping to its 
corresponding nucleotide (`*` for termination).

Note that `addEditedAlleles` also appended several gRNA-level aggregate 
scores to the `GuideSet` object:

```{r}
head(gs)
```

The `score_missense`, `score_nonsense` and `score_silent` columns report 
aggregated scores for each mutation type. They are calculated by summing 
all scores of a given mutation type across the set of edited alleles for 
a given gRNA. The `maxVariant` column indicates the most probable mutation 
type for the given gRNA based on the maximum aggregated score, which is 
stored in `maxVariantScore`. In our example, the highest score for `spacer_4` 
is `score_nonsense`, and so `maxVariant` is set to `nonsense`.



# Session Info

```{r}
sessionInfo()
```


# References

